Remove backward op for 2d depthwise convolution

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70462
* __->__ #70461

Differential Revision: [D33340494](https://our.internmc.facebook.com/intern/diff/D33340494)

jbschlosser committed:
Remove backward op for 2d depthwise convolution

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33340494).

jbschlosser committed:
Update on "Remove backward op for 2d depthwise convolution"


Differential Revision: [D33340494](https://our.internmc.facebook.com/intern/diff/D33340494)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33340494).

ngimel commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -1555,10 +1557,13 @@ std::tuple<Tensor, Tensor, Tensor> convolution_backward(
       std::tie(backend_grad_input, backend_grad_weight) =
-        at::_conv_depthwise2d_backward(grad_output.contiguous(), input.contiguous(), weight, kernel_size,
-            params.stride, params.padding, params.dilation, {output_mask[0], output_mask[1]});
+        conv_depthwise2d_backward_stub(input.device().type(), grad_output.contiguous(), input.contiguous(),
```
I'm not sure this will correctly set device guard and correctly check that inputs are on the same device (and generally, you should not use stubs in functions that are `CompositeExplicitAutograd` or `CompositeImplicitAutograd`) - as `CompositeExplicitAutograd` will not set device guard, and neither will `conv_depthwise2d_backward_stub` call itself.

jbschlosser replied:
Hm good point - any suggestions on the right approach here?
* Keep the stub for now and add manual logic to set device guard & verify inputs are on the same device.
* Avoid the stub and add `convolution_cpu` / `convolution_cuda` with proper dispatch keys, splitting the backend routing logic between the two. This was the longer term plan but it might be better go straight here.
* Some other way to dispatch to the CUDA logic?

ngimel replied:
btw I've noticed one of the landed PRs (for depthwise conv I think?) is already using dispatch in CompositeExplicitAutograd, and thus would break for some combination of devices. 
The second option for handling this looks better. 
In general, Composite*Autograd functions should not use dispatch stub, see #40564. Can we have a test to prevent this from happening? I don't see how this test could look like, cc @bdhirsh, @zou3519 for ideas.

bdhirsh replied:
ah darn, I forgot about the device guard piece :(

I have this crappy regex test from ~6 months ago (https://github.com/pytorch/pytorch/pull/61134) that might work.

I just tried running it though, and it won't easily catch this example without altering the regex in a way that gives a bunch of false positives. We have the native_function `convolution_backward` calling the stub `conv_depthwise2d_backward_stub`. So a simple regex like `op_name in "DEFINE_DISPATCH\((.*)_stub\)"` won't catch it.

ngimel replied:
FYI here's another example where CompositeExplicitAutograd function used dispatch macro and didn't set device guards #70406, but call chain here is sufficiently convoluted also, so that regex probably wouldn't work (and I just manually added device guard, instead of doing the right thing)

jbschlosser replied:
@ngimel @bdhirsh what you do you two think about a workaround for the purposes of this PR that looks like this?
```
dispatch:
    CompositeExplicitAutograd, CUDA: convolution_backward
```

I think this should set the device guard properly for the CUDA case, no?

bdhirsh replied:
talked offline - we agreed that that should work (and will get replaced by a `convolution_backward_cuda` kernel in the longterm)

jbschlosser committed:
Update on "Remove backward op for 2d depthwise convolution"


Differential Revision: [D33340494](https://our.internmc.facebook.com/intern/diff/D33340494)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33340494).

jbschlosser committed:
Update on "Remove backward op for 2d depthwise convolution"


Differential Revision: [D33340494](https://our.internmc.facebook.com/intern/diff/D33340494)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33340494).

bdhirsh reviewed (approved):
lgtm!
